add docs for v5.1: reference -> architecture (#236)
tikv/website: content/docs/5.1/reference/architecture/overview.md

---

This page walks you through the overview of TiKV architecture.

## System architecture

The overall architecture of TiKV is illustrated below:

{{< figure
    src="/img/tikv-architecture.png"
    caption="The architecture of TiKV"
    alt="TiKV architecture diagram" >}}

A TiKV cluster consists of 2 components:

1. [A TiKV Cluster](./#tikv-cluster), which is responsible for storing key-value pair data.
2. [A Placement Driver (PD) Cluster](./#pd-cluster), which works as the manager in a TiKV cluster.

TiKV Clients interact with PD and TiKV through gRPC.

## TiKV Cluster

TiKV stores data in RocksDB, which is a persistent key-value store for fast storage environment. This [article](/deep-dive/key-value-engine/rocksdb/) explains why RocksDB is selected.

TiKV replicates data to multiple machines via [Raft](/deep-dive/consensus-algorithm/raft/) in case of machine failure. Data is written through the interface of Raft instead of to RocksDB. With the implementation of Raft, TiKV becomes a distributed Key-Value storage. Even with a few machine failures, TiKV can automatically complete replicas by virtue of the native Raft protocol, which does not impact the application.

> Based on the Raft layer, TiKV provides two APIs that clients can use to interact with it:

Reviewer: ```suggestion
Based on the Raft layer, TiKV provides two APIs that clients can interact with:
```

---
title: Overview
description: Core concepts and architecture behind TiKV.
menu:
    "5.1":
        parent: Architecture
        weight: 1
---

This page discusses the core concepts and architecture behind TiKV, including:

* The [APIs](#apis) that applications can use to interact with TiKV
* The basic [system architecture](#system) underlying TiKV
* The anatomy of each [instance](#instance) in a TiKV installation

> * The role of core system components, including the [Placement Driver](#placement-driver), [Store](#store), [Region](#region), and [Node](#node)

Reviewer 2: Do you mean `the terminology of the TiKV cluster`

Reviewer 2: Since, region is not a "role", it's a terminology of our own.

Author: yes

Author: fixed

* TiKV's [transaction model](#transactions)
* The role of the [Raft consensus algorithm](#raft) in TiKV
* The [origins](#origins) of TiKV

## APIs

TiKV provides two APIs that you can use to interact with it:

| API           | Description                                                                           | Atomicity     | Use when...                                                                                                            |
|:------------- |:------------------------------------------------------------------------------------- |:------------- |:---------------------------------------------------------------------------------------------------------------------- |

> | Raw           | A lower-level key-value API for interacting directly with individual key-value pairs. | Single key    | Your application requires low latency instead of distributed transactions or multi-version concurrency control (MVCC). |

Reviewer 3: ```suggestion
| Raw           | A lower-level key-value API for interacting directly with individual key-value pairs. | Single key    | Your application requires low latency and doesn't use multi-key transactions. |
```

> | Transactional | A higher-level key-value API that provides ACID semantics                             | Multiple keys | Your application requires distributed transactions and/or MVCC.                                                        |

Reviewer 3: ```suggestion
| Transactional | A higher-level key-value API that provides snapshot isolation transaction                             | Multiple keys | Your application requires distributed transactions.                                                        |
```

## System architecture {#system}

The overall architecture of TiKV is illustrated below:

{{< figure
    src="/img/basic-architecture.png"
    caption="The architecture of TiKV"
    alt="TiKV architecture diagram"
    width="70" >}}

## TiKV instance {#instance}

The architecture of each TiKV instance is illustrated below:

{{< figure
    src="/img/tikv-instance.png"
    caption="TiKV instance architecture"
    alt="TiKV instance architecture diagram"
    width="60" >}}

## Placement driver (PD) {#placement-driver}

> The TiKV placement driver is the cluster manager of TiKV, which periodically checks replication constraints to balance load and data automatically across nodes and regions in a process called **auto-sharding**.

Reviewer 3: ```suggestion
The TiKV placement driver is the cluster manager of TiKV, which periodically and automatically balances workload and data storage across nodes by moving Regions. This process is called **auto-sharding**.
```

## Store

> There is a [RocksDB](https://rocksdb.org) database within each **Store** and it stores data into the local disk.

Reviewer 3: ```suggestion
The **Store** is a TiKV instance. A Store stores multiple peers.
```

move below peer section.

Author: what about

```
A **Store** refers to the storage node in the TiKV cluster (an instance of tikv-server). Each store has a corresponding TiKV instance.
```

> ## Region

Reviewer 3: May be better to explain peer after this.

Author: ```
## Peer

A replica of a Region is called a peer. Multiple peers of the same Region replicate data via the Raft consensus algorithm, so peers are also members of a Raft instance.
```

> **Region** is the basic unit of key-value data movement. Each Region is replicated to multiple Nodes. These multiple replicas form a Raft group.

Reviewer 3: ```suggestion
TiKV shards continuous ranges of keys into **Regions**, and replicates **Regions** via the Raft protocol. When data size increases until reaching a threshold, a Region will be split into multiple. Conversely, if the size of the Region shrinks due to data deletion, two adjacent Regions can be merged into one.
```

## Node

> A TiKV **Node** is just a physical node in the cluster, which could be a virtual machine, a container, etc. Within each Node, there are one or more stores. The data in each Store is split across multiple regions. Data is distributed across Regions using the Raft algorithm.

Reviewer 3: ```suggestion
A TiKV **Store** is a physical node in the cluster, which could be a virtual machine, a container, etc. Within each Node, there can be one or more stores.
```

> When a Node starts, the metadata for the Node, Store, and Region is recorded into the Placement Driver. The status of each Region and Store is regularly reported to the PD.

Reviewer 3: ```suggestion
The status for the Node, Store, and Region will be regularly reported to the Placement Driver.
```

> ## Transaction model {#transactions}

Reviewer 3: ```suggestion
## Transaction {#transactions}
```

> TiKV's transaction model is similar to that of Google's [Percolator](https://ai.google/research/pubs/pub36726), a system built for processing updates to large data sets. Percolator uses an incremental update model in place of a batch-based model.
> 
> TiKV's transaction model provides:
> 
> * **Snapshot isolation** with lock, with semantics analogous to `SELECT ... FOR UPDATE` in SQL
> * Externally consistent reads and writes in distributed transactions

Reviewer 3: ```suggestion
TiKV provides transactions that allow you to read and write across any keys with **Snapshot Isolation** regardless of the physical placement of the Region. TiKV also provides the pessimistic transactions that are semantically analogous to `SELECT ... FOR UPDATE` in SQL.

The underlying transaction model is similar to Google's [Percolator](https://ai.google/research/pubs/pub36726), a system built for processing updates to large data sets.
```

---
title: Overview
description: Core concepts and architecture behind TiKV.
menu:
    "5.1":
        parent: Architecture
        weight: 1
---

This page discusses the core concepts and architecture behind TiKV, including:

* The [APIs](#apis) that applications can use to interact with TiKV
* The basic [system architecture](#system) underlying TiKV
* The anatomy of each [instance](#instance) in a TiKV installation
* TiKV's terminology including [Placement Driver](#placement-driver), [Store](#store), [Region](#region), and [Node](#node)
* TiKV's [transaction model](#transactions)
* The role of the [Raft consensus algorithm](#raft) in TiKV

> * The [origins](#origins) of TiKV

Author: +1

> * The [origins](#origins) of TiKV

Author: PTAL

|:------------- |:------------------------------------------------------------------------------------- |:------------- |:----------------------------------------------------------------------------- |
| Raw           | A lower-level key-value API for interacting directly with individual key-value pairs. | Single key    | Your application requires low latency and doesn't use multi-key transactions. |
| Transactional | A higher-level key-value API that provides snapshot isolation transaction.            | Multiple keys | Your application requires distributed transactions.                           |

## System architecture

The overall architecture of TiKV is illustrated below:

{{< figure
    src="/img/basic-architecture.png"
    caption="The architecture of TiKV"
    alt="TiKV architecture diagram"
    width="70" >}}

## TiKV instance

The architecture of each TiKV instance is illustrated below:

{{< figure
    src="/img/tikv-instance.png"
    caption="TiKV instance architecture"
    alt="TiKV instance architecture diagram"
    width="60" >}}

## Placement driver (PD)

> The TiKV placement driver is the cluster manager of TiKV, which periodically and automatically balances workload and data storage across nodes by moving Regions. This process is called **auto-sharding**.

Author: +1

## Store

> A **Store** refers to the storage node in the TiKV cluster (an instance of tikv-server). Each store has a corresponding TiKV instance.

Reviewer 3: This is ambiguous to the `Node` section: Is node a tikv instance or the machine that runs the tikv?

Author: what about 
```
A **Store** is an instance of tikv-server.
```

Reviewer 3: Modified a little bit:

A **Store** is an instance of a TiKV server in which multiple peers are stored.

## Region

TiKV shards continuous ranges of keys into **Regions**, and replicates **Regions** via the Raft protocol. When data size increases until reaching a threshold, a Region will be split into multiple. Conversely, if the size of the Region shrinks due to data deletion, two adjacent Regions can be merged into one.

## Peer

> A replica of a Region is called a peer. Multiple peers of the same Region replicate data via the Raft consensus algorithm, so peers are also members of a Raft instance.

Reviewer 3: ```suggestion
A replica of a Region is called a peer. Multiple peers of the same Region replicate data via the Raft consensus algorithm.
```

## Node

> A TiKV **Node** is a physical node in the cluster, which could be a virtual machine, a container, etc. Within each Node, there can be one or more stores.

Author: No, Store is the tikv-server process.
